query/src/lib/mutation.ts

```diff
@@ -0,0 +1,113 @@
+import { inject, Injectable, InjectionToken, Signal } from '@angular/core';
+import { injectQueryClient } from './query-client';
+import {
+  DefaultError,
+  MutationObserver,
+  MutationObserverOptions,
+  MutationObserverResult,
+  notifyManager,
+} from '@tanstack/query-core';
+import { firstValueFrom, isObservable, Observable, shareReplay } from 'rxjs';
+import { toSignal } from '@angular/core/rxjs-interop';
+
+type CreateMutationOptions<
+  TData = unknown,
+  TError = DefaultError,
+  TVariables = void,
+  TContext = unknown
+> = Omit<
+  MutationObserverOptions<TData, TError, TVariables, TContext>,
+  'mutationFn'
+> & {
+  mutationFn: (variables: TVariables) => Promise<TData> | Observable<TData>;
+};
+
+type MutationResult<
+  TData = unknown,
+  TError = DefaultError,
+  TVariables = void,
+  TContext = unknown
+> = {
+  mutate: MutationObserver<TData, TError, TVariables, TContext>['mutate'];
+  reset: MutationObserver<TData, TError, TVariables, TContext>['reset'];
+  setOptions: MutationObserver<
+    TData,
+    TError,
+    TVariables,
+    TContext
+  >['setOptions'];
+  result$: Observable<
+    MutationObserverResult<TData, TError, TVariables, TContext>
+  >;
+  toSignal: () => Signal<MutationObserverResult<TData, TError, TVariables, TContext> | undefined>
```

Comment:
Why here we have | undefined and it the result$ version not?

Reply:
`toSignal(...)` requires you to add `undefined` when either `requireSync` or `initialValue` isn't set.
And i havent set either of them, we could give the user the option to set these options on `toSignal(...)` .

Reply:
If we do this though, we should consider adding `toSignalOptions` library wide, since every `toSignal` has an `| undefined` to it, if one or another isn't set.

Reply:
but in query you will always have an initial value. I did not need it there. Here, if you read the signal before calling mutate you will get undefined, right?

Reply:
As far as i know, yes.

Reply:
And in case of an observable it will not fire until it has a value. If that's the case, the code make sense

Reply:
I would suggest, since `observer.next(mutationObserver.getCurrentResult());` is called directly after subscribing to it, that we also set `requireSync` to `true`. The result the `mutationObserver` produces upon subscribing would be an `MutationObserverIdleResult` result and thus should not give any runtime errors. Then i also can remove the `undefined` from it.

Reply:
sounds good

Reply:
@NetanelBasal we probably also should add the `requireSync: true` to `isFetching`, `isMutating` and the `base-query`, the latter one reveals that when you add the typings mentioned in #122 that indeed the return type would be also `| undefined` 

![grafik](https://github.com/ngneat/query/assets/8077014/8443e4c2-cdbe-4d83-b816-e90ec61492d7)

![grafik](https://github.com/ngneat/query/assets/8077014/694ad537-5de8-4438-91e1-4479a0167e29)

So just to mitigate any problems later, we maybe should add a) the typings to the observable and b) `requireSync: true` to it